sled-agent/src/opte.rs

```diff
@@ -0,0 +1,334 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Interactions with the Oxide Packet Transformation Engine (OPTE)
+
+use crate::illumos::addrobj;
+use crate::illumos::addrobj::AddrObject;
+use crate::illumos::dladm;
+use crate::illumos::dladm::Dladm;
+use crate::illumos::dladm::PhysicalLink;
+use crate::illumos::vnic::Vnic;
+use crate::illumos::zone;
+use crate::illumos::zone::Zones;
+use ipnetwork::IpNetwork;
+use macaddr::MacAddr6;
+use opte::api::IpCidr;
+use opte::api::Ipv4Cidr;
+use opte::api::Ipv4PrefixLen;
+use opte::api::MacAddr;
+pub use opte::engine::geneve::Vni;
+use opte::oxide_vpc::api::AddRouterEntryIpv4Req;
+use opte::oxide_vpc::api::PhysNet;
+use opte::oxide_vpc::api::RouterTarget;
+use opte::oxide_vpc::api::SetVirt2PhysReq;
+use opte_ioctl::OpteHdl;
+use slog::Logger;
+use std::net::IpAddr;
+use std::net::Ipv6Addr;
+use std::sync::atomic::AtomicU64;
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+use uuid::Uuid;
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error(transparent)]
+    Opte(#[from] opte_ioctl::Error),
+
+    #[error(transparent)]
+    Dladm(#[from] dladm::Error),
+
+    #[error(transparent)]
+    Zone(#[from] zone::Error),
+
+    #[error(transparent)]
+    AddrObj(#[from] addrobj::Error),
+}
```

Comment:
As a heads up, I did an overhaul of error handling here: https://github.com/oxidecomputer/omicron/pull/967

I expect this will conflict, for one of us. My guiding principle was: "any returned error from the callee should be providing enough context that the caller can immediately return it, without re-adding the arguments again".

Especially for these new types of `Error::Opte`, we may want to provide an additional context message in the error, so we can be more informative about where things went wrong.

Reply:
Thanks for the heads up! Realistically, I won't get to merge this for a bit still, so it's likely that you'll merge #967 first. I'll merge that into this PR, in that case.